File: src/Solutions/Euler109.hs
module Solutions.Euler109 where
import Control.Monad

data Dart = Single Int | Double Int | Triple Int deriving (Eq, Ord, Show)
type Throw = [Dart]
dartVal (Single x) = x
dartVal (Double x) = 2 * x
dartVal (Triple x) = 3 * x
isDouble throw = case last throw of
    Double _ -> True
    _ -> False

darts = [Single 25, Double 25] ++ ([Single, Double, Triple] <*> [1..20])

checkoutWith n = length [() |
    i <- [1..3],
    throw <- replicateM i darts,
    i /= 3 || head throw >= head (tail throw),
    sum (map dartVal throw) == n,
    isDouble throw]

checkoutWithUpTo n = sum $ map checkoutWith [2..n]

euler109 :: IO String
euler109 = return $ show $ checkoutWithUpTo 99

File: rust/src/solutions/euler109.rs
use itertools::Itertools;

#[derive(Copy, Clone, Ord, PartialOrd, Eq, PartialEq)]
enum Dart {
    Single(u32),
    Double(u32),
    Triple(u32),
}

impl Dart {
    fn value(&self) -> u32 {
        match self {
            Dart::Single(d) => *d,
            Dart::Double(d) => 2 * *d,
            Dart::Triple(d) => 3 * *d,
        }
    }
}

fn all_darts() -> Vec<Dart> {
    let mut all_darts = vec![Dart::Single(25), Dart::Double(25)];
    for d in 1..=20 {
        all_darts.extend(vec![Dart::Single(d), Dart::Double(d), Dart::Triple(d)]);
    }
    all_darts
}

fn check_out_up_to(n: u32) -> usize {
    (1..=3)
        .flat_map(|i| {
            (0..i)
                .map(|_| all_darts())
                .multi_cartesian_product()
                .filter(move |throws| {
                    (i < 3 || throws[0] >= throws[1])
                        && throws.iter().map(|throw| throw.value()).sum::<u32>() <= n
                        && throws.last().map_or(false, |throw| match throw {
                            Dart::Double(_) => true,
                            _ => false,
                        })
                })
        })
        .count()
}

pub fn euler109() -> String {
    check_out_up_to(99).to_string()
}
